parser.y

```diff
@@ -6022,45 +6033,26 @@ SelectStmtFromDualTable:
 	}
 
 SelectStmtFromTable:
-	SelectStmtBasic "FROM"
-	TableRefsClause WhereClauseOptional SelectStmtGroup HavingClause WindowClauseOptional
+	SelectStmtBasic FromClauseOptional WhereClauseOptional SelectStmtGroup HavingClause WindowClauseOptional
 	{
 		st := $1.(*ast.SelectStmt)
-		st.From = $3.(*ast.TableRefsClause)
-		lastField := st.Fields.Fields[len(st.Fields.Fields)-1]
-		if lastField.Expr != nil && lastField.AsName.O == "" {
-			lastEnd := parser.endOffset(&yyS[yypt-5])
-			lastField.SetText(parser.src[lastField.Offset:lastEnd])
-		}
-		if $4 != nil {
-			st.Where = $4.(ast.ExprNode)
-		}
-		if $5 != nil {
-			st.GroupBy = $5.(*ast.GroupByClause)
-		}
-		if $6 != nil {
-			st.Having = $6.(*ast.HavingClause)
-		}
-		if $7 != nil {
-		    st.WindowSpecs = ($7.([]ast.WindowSpec))
+		if $2 != nil {
+			st.From = $2.(*ast.TableRefsClause)
 		}
-		$$ = st
-	}
-
-SelectStmt:
-	SelectStmtBasic OrderByOptional SelectStmtLimit SelectLockOpt
-	{
-		st := $1.(*ast.SelectStmt)
-		st.LockTp = $4.(ast.SelectLockType)
 		lastField := st.Fields.Fields[len(st.Fields.Fields)-1]
 		if lastField.Expr != nil && lastField.AsName.O == "" {
 			src := parser.src
 			var lastEnd int
+			lastEnd = lastField.Offset
 			if $2 != nil {
-				lastEnd = yyS[yypt-2].offset-1
+				lastEnd = yyS[yypt-4].offset-1
 			} else if $3 != nil {
+				lastEnd = yyS[yypt-3].offset-1
+			} else if $4 != nil {
+				lastEnd = yyS[yypt-2].offset-1
+			} else if $5 != nil {
 				lastEnd = yyS[yypt-1].offset-1
-			} else if $4 != ast.SelectLockNone {
+			} else if $6 != nil {
 				lastEnd = yyS[yypt].offset-1
 			} else {
 				lastEnd = len(src)
```

Comment:
Here the problem is the assumption: there is only one blank character between productions. In fact, there are cases like 
```
select        ((a+1))                   from t
```

```
FAIL: session_test.go:1250: testSessionSuite.TestFieldText

session_test.go:1273:
    c.Assert(result.Fields()[0].ColumnAsName.O, Equals, tt.field)
... obtained string = "((a+1))    "
... expected string = "((a+1))"
```

To calculate the last offset of a Production, you can use `lastEnd := parser.endOffset(&yyS[yypt-4])`, which handles this case elegantly.